tests/MongoDB.Driver.Core.Tests/Core/Connections/BinaryConnectionTests.cs

```diff
@@ -212,6 +212,48 @@ public void Dispose_should_raise_the_correct_events()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void ReceiveMessage_should_throw_a_FormatException_when_message_is_an_invalid_size(
+            [Values(-1, 48000001)]
+            int length,
+            [Values(false, true)]
+            bool async)
+        {
+            using (var stream = new BlockingMemoryStream())
+            {
+                var bytes = BitConverter.GetBytes(length);
+                stream.Write(bytes, 0, bytes.Length);
+                stream.Seek(0, SeekOrigin.Begin);
+                var encoderSelector = new ReplyMessageEncoderSelector<BsonDocument>(BsonDocumentSerializer.Instance);
+
+                Exception exception;
+                if (async)
+                {
+                    _mockStreamFactory
+                        .Setup(f => f.CreateStreamAsync(_endPoint, CancellationToken.None))
+                        .ReturnsAsync(stream);
+                    _subject.OpenAsync(CancellationToken.None).GetAwaiter().GetResult();
+                    exception = Record
+                        .ExceptionAsync(async () => await _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None))
+                        .GetAwaiter()
+                        .GetResult();
```

Comment:
If we're going to use `.GetAwaiter().GetResult()` anyway I don't see any advantage to using `ExceptionAsync` (a search reveals that we *have* used `ExceptionAsync` before but in very few places).

Lines 237-240 could just be:

    exception = Record.Exception(() => _subject.ReceiveMessageAsync(10, encoderSelector, _messageEncoderSettings, CancellationToken.None).GetAwaiter().GetResult());

Reply:
Since we are choosing not to make the test method `async` and are thus unable to `await Record.ExceptionAsync`, then I agree that the above refactoring is perfectly reasonable.

Playing devil's advocate though: Is there any *disadvantage* to using the current format though? Your previous comment alluded to finding hits on Google regarding issues if we block during an async test method, which shouldn't be an issue here as the method is declared as a synchronous test method.  IMHO, one minor advantage of the `async () => await...` format is that it makes it exceedingly clear that the method we're testing is an `async` method (and it's "front loaded" onto the statement as opposed to ".GetAwaiter().GetResult()` which happens later).

Reply:
Playing devil's advocate though: is there any **advantage** to using `ExceptionAsync`? I would say no, because the overall test method is not an async method, so `ExceptionAsync` is kind of out of place.

Using `ExceptionAsync` also does NOT avoid the need for using `GetAwaiter().GetResult()`. It just forces you to move the `GetAwaiter().GetResult()` to *outside* of the `ExceptionAsync`.

I'd rather call `GetAwaiter().GetResult()` on the original async call than on the `Task<Exception>` task returned by `ExecuteAsync`.

I agree with the comment that it's nice to use `await`, which is why we were tempted to start using async test methods. But in this case, the await is both unnecessary (because no work is done after the await) and at the wrong level (it's nested inside a lambda and not part of the actual test method).